src/libraries/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs

```diff
@@ -32,8 +32,15 @@ public static class CollectionExtensions
 
             if (!dictionary.ContainsKey(key))
             {
-                dictionary.Add(key, value);
-                return true;
+                try
+                {
+                    dictionary.Add(key, value);
+                    return true;
+                }
+                catch (ArgumentException)
+                {
+                    // this exception will be raised if an entry with the same key is added by another thread between the contains key and the add
```

Comment:
I do not think that this is a good change.

Instance methods in .NET class libraries are not thread safe by default. It is up to the calling code to ensure that the collection is not mutated in parallel from multiple threads.

Also, there is no guarantee that this exception was caused by thread safety issue. This try/catch can be just masking unrelated problem.

Reply:
thanks for the feedback. I thought about testing the exception's message as well to pinpoint to this specific case but I was worried about performance impact. Any thoughts on that specifically?

I do believe the concern is still legitimate. In my scenario I'm using a `ConcurrentDictionary` as the entry point of my code gets iterated over in a parallel for each. I think that calling `ConcurrentDictionary.TryAdd` expecting it "to be safe" is a fair expectation.

Reply:
> I thought about testing the exception's message as well to pinpoint to this specific case but I was worried about performance impact. Any thoughts on that specifically?

This would not work, e.g. the exception messages can be localized.

If you are using ConcurrentDictionary and depend on its thread safety guarantees, you should be calling ConcurrentDictionary methods directly and not use these extension methods.

Reply:
>  I think that calling ConcurrentDictionary.TryAdd expecting it "to be safe" is a fair expectation.

It is not fair to expect that instance methods are thread safe by default, unless the documentation explicitly says so.

Reply:
If the type is declared to be thread-safe, my expectation would be that all public methods are thread-safe as well. `ConcurrentDictionary` is documented as thread-safe:

> Represents a thread-safe collection of key/value pairs that can be accessed by multiple threads concurrently.

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-5.0

And the documentation here states that there is no need for additional sync in user code.

> The .NET Framework 4 introduces the System.Collections.Concurrent namespace, which includes several collection classes that are both thread-safe and scalable. Multiple threads can safely and efficiently add or remove items from these collections, without requiring additional synchronization in user code.

https://docs.microsoft.com/en-us/dotnet/standard/collections/thread-safe/

Reply:
The code changed here is not part of ConcurrentDictionary that is documented be thread safe.

The code changed here is an extension method for `IDictionary` interface. The ConcurrentDictionary thread safety guarantees do not apply to `IDictionary`.

As I have said - if you are using ConcurrentDictionary and depend on its thread safety guarantees, you should be calling ConcurrentDictionary methods directly and not use these extension methods.